### ISSUE TITLE
NVIDIA nv0 CUDA memory allocation fails during train() preparation with quantized model

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#641]
Earliest public failure is `recipe.train().run()` reaching `Gpu::allocate` (`recipe.rs` ~6360-6387) through the NVIDIA `Driver::Cuda` path (`driver.allocate` -> `Gpu::status` -> `RecipeError`), producing `device nv0 Nvidia: Nvidia allocation failed: 2`. The inspected implementation path is `Gpu::allocate` called from `upload` (`recipe.rs` 6412) or `native_program` (`recipe.rs` 6268) during training setup before any epoch arithmetic. The model `recipe.model().conv(8, 1).leak().norm(batch).iq(1).s.svm().cos().qi(5).k.s.loss(mae)` with `.iq(1).s` and `.qi(5).k.s` quantization reaches the same `Driver::Cuda` allocation check; the temporal SQLite loader is loaded but does not alter the earliest cause. No existing issue represents this CUDA allocation failure path.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `7ee57f1a2f6d5450`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{
  "verdict": "new",
  "issue": 0,
  "title": "NVIDIA nv0 CUDA memory allocation fails during train() preparation with quantized model",
  "body": "Earliest public failure is `recipe.train().run()` reaching `Gpu::allocate` (`recipe.rs` ~6360-6387) through the NVIDIA `Driver::Cuda` path (`driver.allocate` -> `Gpu::status` -> `RecipeError`), producing `device nv0 Nvidia: Nvidia allocation failed: 2`. The inspected implementation path is `Gpu::allocate` called from `upload` (`recipe.rs` 6412) or `native_program` (`recipe.rs` 6268) during training setup before any epoch arithmetic. The model `recipe.model().conv(8, 1).leak().norm(batch).iq(1).s.svm().cos().qi(5).k.s.loss(mae)` with `.iq(1).s` and `.qi(5).k.s` quantization reaches the same `Driver::Cuda` allocation check; the temporal SQLite loader is loaded but does not alter the earliest cause. No existing issue represents this CUDA allocation failure path.",
  "rationale": "Searched open nv0 issues (346, 506, 502, 497, 345, 325, 324, 317, 315, 293, 282, 271, and others) and the full repository; none cover a CUDA memory allocation failure (`Nvidia allocation failed: 2`) through `Gpu::allocate` (`recipe.rs` 6375). Existing nv0 issues address native artifact compilation stalls (`compile_native_artifact`), performance epoch stalls, or predictor/scan compilation delays. This packet reaches the `Driver::Cuda` `allocate` path (not `compile_native_artifact`) with a deterministic CUDA memory error, a materially different earliest operation and implementation path. The quantized model blocks (`.iq(1).s`, `.qi(5).k.s`) and dataset are additional evidence for the same loader-ignored path, not a different defect."
}
```

## Deterministic failure packet

id=7ee57f1a2f6d5450
base=commit=10148e25421cc98c32cf7fbe59237263a1b98f73 tracked_tree=clean
cursor=seed:0 cursor:18761 next:18762 step:53196099145139 composition:21668088059659
data=path:"data/temporal/series_sqlite.sqlite" hash:8adec46e1a4cb73b
configuration=data_mode:3 model:recipe.model().conv(8, 1).leak().norm(batch).iq(1).s.svm().cos().qi(5).k.s.loss(mae) loss:3 arithmetic:9 stop:1 rate:0 lifecycle:0
expected=training, optional resume, and inference produce finite numerical results through the public Recipe API
observed=phase:training message:device nv0 Nvidia: Nvidia allocation failed: 2
output=phase:training message:device nv0 Nvidia: Nvidia allocation failed: 2
replay=phase:training message:device nv0 Nvidia: Nvidia allocation failed: 2 stable:true
command=RECIPE_DEVICE=nv0 cargo run --bin recipe -- /tmp/recipe-composition-repro.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = "/tmp/recipe-composition-repro.ogdl";
	let data = recipe.data("data/temporal/series_sqlite.sqlite")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.conv(8, 1).leak().norm(batch).iq(1).s
		.svm().cos().qi(5).k.s
		.loss(mae);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(21668088059659)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.f(6, 9)
		.save(bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=nv0

**Comment by nm-z** [comment]
Packet 4ee6a0ce287a2f83: phase=training; device nv1; observed `device nv1 Nvidia: Nvidia allocation failed: 2`; reproduction reaches `recipe.train().optimizer(adamw).lr(0.001).seed(17).epochs(20).log(all).fp(64).save(&bundle).run(&model, &data)` before any epoch arithmetic and panics at recipe.rs:21854. Source inspection (`recipe.rs` 16033 `Gpu::allocate`, 16046-16048 `Driver::Cuda` allocate path, 15857-15859 `self.status(..., "allocation")`) confirms the same `Gpu::allocate` -> `Driver::Cuda` -> `RecipeError` path that produces `device {name} Nvidia: Nvidia allocation failed: 2`. Issue #641 records the identical error message and the same earliest operation (`recipe.train().run()` during training setup before epochs) with the same confirmed `Gpu::allocate` / `Driver::Cuda` implementation path. Differences in device label (nv1 vs nv0), model composition (`conv/selu/rnn/qi/k.s.moe/lgbm` vs quantized model), dataset (`data/numeric/sample_subfolders`), and seed are configuration reached after the same failure point; classify by earliest operation/path, not by device/model. No hang or timeout evidence: the packet is an immediate allocation panic with no `do_wait`/process-state record and no progress (no advancing epochs or growing output), so it is not a performance/hang issue. Verdict: new evidence for existing closed issue, comment.

## Machine provenance

- Discoverer: Recipe composition traversal
- Failure fingerprint: `4ee6a0ce287a2f83`
- Recipe source commit: `560addd67cbe5708d36d91ae8d37a88751ba716a`
- Embedded Recipe `recipe.rs` blob: `e149eb47c95b7992c006a1a6a8f18616b5e07a85`
- Source identity: The packet captures the exact working-tree Recipe source independently of the Recipe Machine commit.
- Classifier provider: `dsh`
- Classifier model: `openrouter/thinkingmachines/inkling:free`
- Classifier effort: `default`

### Classifier decision

```json
{"verdict":"comment","issue":641,"title":"NVIDIA nv1 CUDA memory allocation fails during train() preparation","body":"Packet 4ee6a0ce287a2f83: phase=training; device nv1; observed `device nv1 Nvidia: Nvidia allocation failed: 2`; reproduction reaches `recipe.train().optimizer(adamw).lr(0.001).seed(17).epochs(20).log(all).fp(64).save(&bundle).run(&model, &data)` before any epoch arithmetic and panics at recipe.rs:21854. Source inspection (`recipe.rs` 16033 `Gpu::allocate`, 16046-16048 `Driver::Cuda` allocate path, 15857-15859 `self.status(..., \"allocation\")`) confirms the same `Gpu::allocate` -> `Driver::Cuda` -> `RecipeError` path that produces `device {name} Nvidia: Nvidia allocation failed: 2`. Issue #641 records the identical error message and the same earliest operation (`recipe.train().run()` during training setup before epochs) with the same confirmed `Gpu::allocate` / `Driver::Cuda` implementation path. Differences in device label (nv1 vs nv0), model composition (`conv/selu/rnn/qi/k.s.moe/lgbm` vs quantized model), dataset (`data/numeric/sample_subfolders`), and seed are configuration reached after the same failure point; classify by earliest operation/path, not by device/model. No hang or timeout evidence: the packet is an immediate allocation panic with no `do_wait`/process-state record and no progress (no advancing epochs or growing output), so it is not a performance/hang issue. Verdict: new evidence for existing closed issue, comment.","rationale":"Earliest failing public operation is `recipe.train().run()`; confirmed implementation path is `Gpu::allocate` (`recipe.rs` 16033) through NVIDIA `Driver::Cuda` (`recipe.rs` 16046-16048) via `self.status(..., \"allocation\")` (`recipe.rs` 15857-15859), producing `Nvidia allocation failed: 2`. Issue #641 describes the same message (`device nv0 Nvidia: Nvidia allocation failed: 2`), same `Gpu::allocate` / `upload` or `native_program` training-setup path, and same earliest operation. nv1 vs nv0 and different model/dataset/seed are post-failure configuration; instructions say different reproductions are duplicates when they stop at the same public operation with the same observed failure or confirmed cause. No source evidence supports a different earliest cause (e.g., different phase/function). Packet shows no progress or wait-state evidence, so it is not a timeout/hang. Therefore comment on #641."}
```

## Deterministic failure packet

id=4ee6a0ce287a2f83
base=commit=560addd67cbe5708d36d91ae8d37a88751ba716a recipe_source=e149eb47c95b7992c006a1a6a8f18616b5e07a85
cursor=cursor:20187 next:20188 composition:20187
data=path:data/numeric/sample_subfolders
configuration=cursor:20187 seed:17
expected=the generated public Recipe composition trains and infers with finite values
observed=phase:training message:device nv1 Nvidia: Nvidia allocation failed: 2
output=phase:training
thread 'main' (3344433) panicked at recipe.rs:21854:13:
device nv1 Nvidia: Nvidia allocation failed: 2
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
replay=phase:training message:device nv1 Nvidia: Nvidia allocation failed: 2 stable:true
command=RECIPE_DEVICE=nv1 target/debug/recipe test.rs
reproduction:
```rust
use recipe::*;
use std::path::{Path, PathBuf};

fn phase(value: &str) {
    if let Some(path) = std::env::var_os("RECIPE_COMPOSITION_PHASE_PATH") {
        std::fs::write(path, value).expect("cannot write phase");
    }
}

fn width(path: &Path) -> usize {
    let text = std::fs::read_to_string(path).expect("cannot read saved model");
	let mut shape = text.lines().find_map(|line| line.trim().strip_prefix("shape ")).expect("saved model has no shape").split_whitespace();
    shape.next().expect("saved model has no channels").parse::<usize>().expect("saved channels are invalid") * shape.next().expect("saved model has no length").parse::<usize>().expect("saved length is invalid")
}

fn main() {
    let root = std::env::var_os("RECIPE_TRIAL_DIRECTORY").map(PathBuf::from).unwrap_or_else(std::env::temp_dir);
    let bundle = root.join(format!("recipe-model-{}.ogdl", std::process::id()));
    phase("setup");
    let data = recipe.data("data/numeric/sample_subfolders").target("target").norm(z_score);
    let model = recipe.model().conv(16, 5).selu().rnn(16).qi(3).k.s.moe(2, [perc(24).elu().qi(4).nf, layer(32), rnn(32).tanh()]).lgbm(8).layer(1).loss(bce);
    phase("training");
    let report = recipe.train().optimizer(adamw).lr(0.001).seed(17).epochs(20).log(all).fp(64).save(&bundle).run(&model, &data);
    assert!(report.final_loss().is_finite());
    phase("inference");
    let output = recipe.infer(&bundle, &vec![0.0; width(&bundle)]);
    assert!(!output.is_empty() && output.iter().all(|value| value.is_finite()));
    std::fs::remove_file(bundle).expect("cannot remove saved model");
}
```
backend=nv1
